src/rascal/math/spherical_harmonics.cc

```diff
@@ -0,0 +1,363 @@
+/**
+ * @file   spherical_harmonics.cc
+ *
+ * @author  Felix Musil <felix.musil@epfl.ch>
+ * @author  Max Veit <max.veit@epfl.ch>
+ * @author  Alex
+ *
+ * @date   14 October 2018
+ *
+ * @brief implementation of the spherical harmonics, optimized, with gradients
+ *
+ * Copyright  2018  Felix Musil, Max Veit, COSMO (EPFL), LAMMM (EPFL)
+ *
+ * Rascal is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3, or (at
+ * your option) any later version.
+ *
+ * Rascal is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this software; see the file LICENSE. If not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+#include "rascal/math/spherical_harmonics.hh"
+
+#include <iostream>
+
+using namespace rascal::math;  // NOLINT
```

Comment:
Can we get rid of this `using namespace`? Or rather, if the alternative is to wrap the entire file in `namespace rascal::math`, then why does our style guide discourage `using namespace` statements for implementation files?

Reply:
>  then why does our style guide discourage using namespace statements for implementation files?

The usual rational is to forbid `using namespace xxx` in headers files, since this introduces namespace pollution in every file including the header. For implementation / .cc files, this is usually fine, since the `using namespace xxx` is scoped to this one and only file.

As far as the style guide goes, I have no idea.

Reply:
Ok then, this usage seems sensible enough to me, so I’d suggest updating the style guide if it doesn’t already allow it (we might need to do this in the next group meeting). If possible, I’d also like to update the clang-format file to allow using-namespace declarations in implementation files; those `NOLINT` directives are super ugly.

Reply:
> those NOLINT directives are super ugly.

This is a limitation of cpplint. The rule is either enabled or disabled for all files in a folder, and I could not find a way to separate files by extension. We could have separated folders for .cc and .hh files, with separated CPPLINT.cfg, but I remember this was not the preferred solution when we discussed #189.

Reply:
Ugh, I knew it wasn’t going to be that simple. Let’s discuss it at the next group meeting, then.